Using ArrayOf in MediaType→Attributes→Attribute causes generator crash

I have a DSL that causes a nil pointer dereference crash when running goagen. This is not an ideal way of telling me I made a mistake :P. Minimal repro design:

```
package design

import (
	. "github.com/goadesign/goa/design"
	. "github.com/goadesign/goa/design/apidsl"
)

var _ = API("crash", func() {})

var _ = MediaType("thing", func() {
	Attributes(func() {
		Attribute("rec", ArrayOf("thing"))
	})

	View("default", func() {
		Attribute("rec")
	})
})
```

The error in this design is that I should have used CollectionOf, not ArrayOf.

Running `goagen bootstrap` on this design results in a nil pointer dereference crash:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4a37e2]

goroutine 1 [running]:
panic(0x795780, 0xc4200100e0)
	/usr/lib/go/src/runtime/panic.go:500 +0x1a1
github.com/goadesign/goa/design.(*UserTypeDefinition).IsPrimitive(0x0, 0xc420163670)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/types.go:642 +0x22
github.com/goadesign/goa/design.(*AttributeDefinition).Validate(0xc42018be80, 0xc420163680, 0xf, 0x9ebd60, 0xc42018bf00, 0xc420163670)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:451 +0x139
github.com/goadesign/goa/design.(*AttributeDefinition).Validate(0xc42018bf00, 0xc420163670, 0xc, 0x9ebfe0, 0xc420168e80, 0xc420163660)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:484 +0x982
github.com/goadesign/goa/design.(*AttributeDefinition).Validate(0xc42018bd80, 0x0, 0x0, 0x9ebfe0, 0xc420168e80, 0xc41fff4d49)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:479 +0x451
github.com/goadesign/goa/design.(*UserTypeDefinition).Validate(0xc420168e80, 0x0, 0x0, 0x9ebea0, 0xc42016f3c0, 0x40f69e)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:520 +0xa8
github.com/goadesign/goa/design.(*MediaTypeDefinition).Validate(0xc42016f3c0, 0x9f13a0)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:528 +0xa6
github.com/goadesign/goa/design.(*APIDefinition).Validate.func2(0xc42016f3c0, 0xc4201c6510, 0xc4201634c0)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:146 +0x49
github.com/goadesign/goa/design.(*APIDefinition).IterateMediaTypes(0xc4201922c0, 0xc42020f9d8, 0x0, 0x0)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/definitions.go:528 +0x1d8
github.com/goadesign/goa/design.(*APIDefinition).Validate(0xc4201922c0, 0x9ebce0, 0xc4201922c0)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/validation.go:148 +0x333
github.com/goadesign/goa/dslengine.validateSet(0xc420163610, 0x1, 0x1, 0xa2b908, 0x0)
	/home/dave/hack/go/src/github.com/goadesign/goa/dslengine/runner.go:312 +0x120
github.com/goadesign/goa/design.(*APIDefinition).IterateSets(0xc4201922c0, 0x84ca30)
	/home/dave/hack/go/src/github.com/goadesign/goa/design/definitions.go:458 +0x96
github.com/goadesign/goa/dslengine.Run(0x9ed460, 0xc420168ea0)
	/home/dave/hack/go/src/github.com/goadesign/goa/dslengine/runner.go:108 +0x197
main.main()
	/home/dave/hack/go/src/go.universe.tf/gipam/goagen171272070/main.go:24 +0x63
```

--- a/design/apidsl/type.go
+++ b/design/apidsl/type.go
@@ -82,7 +82,11 @@ func ArrayOf(v interface{}, dsl ...func()) *design.Array {
 	t, ok = v.(design.DataType)
 	if !ok {
 		if name, ok := v.(string); ok {
-			t = design.Design.Types[name]
+			if ut, ok := design.Design.Types[name]; ok {
+				t = ut
+			} else if mt, ok := design.Design.MediaTypes[name]; ok {
+				t = mt
+			}
 		}
 	}
 	// never return nil to avoid panics, errors are reported after DSL execution

--- a/design/apidsl/type_test.go
+++ b/design/apidsl/type_test.go
@@ -175,7 +175,7 @@ var _ = Describe("ArrayOf", func() {
 			Ω(dslengine.Errors).ShouldNot(HaveOccurred())
 		})
 
-		It("produces a media type", func() {
+		It("produces a user type", func() {
 			Ω(ar).ShouldNot(BeNil())
 			Ω(ar.TypeName).Should(Equal("names"))
 			Ω(ar.Type).ShouldNot(BeNil())
@@ -188,6 +188,41 @@ var _ = Describe("ArrayOf", func() {
 			Ω(et).ShouldNot(BeNil())
 			Ω(et.Type).Should(BeAssignableToTypeOf(&UserTypeDefinition{}))
 			Ω(et.Type.(*UserTypeDefinition).TypeName).Should(Equal("name"))
+		})
+	})
+
+	Context("defined with a media type name", func() {
+		var mt *MediaTypeDefinition
+		BeforeEach(func() {
+			dslengine.Reset()
+			mt = MediaType("application/vnd.test", func() {
+				Attributes(func() {
+					Attribute("ut", ArrayOf("application/vnd.test"))
+				})
+				View("default", func() {
+					Attribute("ut")
+				})
+			})
+		})
+
+		JustBeforeEach(func() {
+			dslengine.Run()
+			Ω(dslengine.Errors).ShouldNot(HaveOccurred())
+		})
+
+		It("produces a user type", func() {
+			Ω(mt).ShouldNot(BeNil())
+			Ω(mt.TypeName).Should(Equal("Test"))
+			Ω(mt.Type).ShouldNot(BeNil())
+			Ω(mt.Type.ToObject()).ShouldNot(BeNil())
+			Ω(mt.Type.ToObject()).Should(HaveKey("ut"))
+			ut := mt.Type.ToObject()["ut"]
+			Ω(ut.Type).ShouldNot(BeNil())
+			Ω(ut.Type).Should(BeAssignableToTypeOf(&Array{}))
+			et := ut.Type.ToArray().ElemType
+			Ω(et).ShouldNot(BeNil())
+			Ω(et.Type).Should(BeAssignableToTypeOf(&MediaTypeDefinition{}))
+			Ω(et.Type.(*MediaTypeDefinition).TypeName).Should(Equal("Test"))
 		})
 	})
 })

--- a/middleware/security/jwt/jwt_test.go
+++ b/middleware/security/jwt/jwt_test.go
@@ -18,9 +18,7 @@ var _ = Describe("Middleware", func() {
 	var respRecord *httptest.ResponseRecorder
 	var request *http.Request
 	var handler goa.Handler
-	var scopesFetcher func(context.Context) []string
 	var middleware goa.Middleware
-	var requiredScopes []string
 	var dispatchResult error
 	var fetchedToken *jwtpkg.Token
 
@@ -30,16 +28,12 @@ var _ = Describe("Middleware", func() {
 			Name: "Authorization",
 		}
 		respRecord = httptest.NewRecorder()
-		requiredScopes = []string{"scope1"}
 		request, _ = http.NewRequest("GET", "http://example.com/", nil)
 		// HS256 {"scopes":"scope1","admin":true}, signed with "keys"
 		request.Header.Set("Authorization", "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzY29wZXMiOiJzY29wZTEiLCJhZG1pbiI6dHJ1ZX0.UCvEfbD_yuS5dCZidxZgogVi2yF0ZVecMsQQbY1HJy0")
 		handler = func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
 			fetchedToken = jwt.ContextJWT(ctx)
 			return nil
-		}
-		scopesFetcher = func(ctx context.Context) []string {
-			return requiredScopes
 		}
 	})
 

